sematic/graph.py

```diff
@@ -0,0 +1,544 @@
+            cloning the graph. The second element is the list of run IDs whose
+            cloned future's state to reset to CREATED.
+        """
+        skip_run_ids: List[RunID] = []
+
+        reset_run_ids: List[RunID] = [reset_from]
+
+        ancestor_run_ids = self._get_run_ancestor_ids(reset_from)
+        reset_run_ids += ancestor_run_ids
+
+        for ancestor_run_id in [reset_from] + ancestor_run_ids:
+            downstream_run_ids = self._get_run_downstream_ids(ancestor_run_id)
+            reset_run_ids += downstream_run_ids
+
+            for downstream_run_id in downstream_run_ids:
+                downstream_descendant_ids = self._get_run_descendant_ids(
+                    downstream_run_id
+                )
+                skip_run_ids += downstream_descendant_ids
+
+        return skip_run_ids, reset_run_ids
+
+    @memoized_indexed
+    def _get_artifact_value(self, artifact_id: str) -> Any:
+        artifact = self._artifacts_by_id[artifact_id]
+        return get_artifact_value(artifact, self.storage)
+
+    def _get_cloned_future_inputs(
+        self, run_id: RunID, cloned_graph: ClonedFutureGraph
+    ) -> Tuple[Dict[str, Any], Dict[str, Artifact]]:
+        kwargs: Dict[str, Any] = {}
+
+        input_edges = self._edges_by_destination_id[run_id]
+
+        run_input_artifacts: Dict[str, Artifact] = {}
+
+        for input_edge in input_edges:
+            if input_edge.destination_name is None:
+                raise RuntimeError("Input edge misses destination name")
+
+            value = None
+
+            if input_edge.artifact_id is not None:
+                artifact = self._artifacts_by_id[input_edge.artifact_id]
+                run_input_artifacts[input_edge.destination_name] = artifact
+                value = self._get_artifact_value(artifact.id)
+
+            if input_edge.source_run_id is not None:
+                # We set the input as the upstream future to mimick what
+                # happens in a greenfield resolution.
+                kwargs[
+                    input_edge.destination_name
+                ] = cloned_graph.futures_by_original_id[input_edge.source_run_id]
+            elif input_edge.artifact_id is not None:
+                kwargs[input_edge.destination_name] = value
+            else:
+                raise RuntimeError(
+                    "Invalid input edge had no source run or associated artifact"
+                )
+
+        return kwargs, run_input_artifacts
+
+    def _get_run_output(self, run_id: RunID) -> Tuple[Any, Optional[Artifact]]:
+        output_edges = self._edges_by_source_id[run_id]
+
+        for output_edge in output_edges:
+            if output_edge.artifact_id is None:
+                return None, None
+
+            artifact = self._artifacts_by_id[output_edge.artifact_id]
+            value = self._get_artifact_value(artifact.id)
+            return value, artifact
+
+        return None, None
+
+    def _set_cloned_parent_future(
```

Comment:
This doesn't set the parent future to a value directly specified in the arguments.
Please consider renaming to `_update_cloned_parent_future`, or `_initialize_cloned_parent_future`.

Reply:
This method sets the cloned parent future. The method is appropriately named.